src/core/qgsdiagramrenderer.cpp

```diff
@@ -804,6 +851,275 @@ void QgsLinearlyInterpolatedDiagramRenderer::writeXml( QDomElement &layerElem, Q
   layerElem.appendChild( rendererElem );
 }
 
+QgsStackedDiagramRenderer *QgsStackedDiagramRenderer::clone() const
+{
+  return new QgsStackedDiagramRenderer( *this );
+}
+
+QSizeF QgsStackedDiagramRenderer::sizeMapUnits( const QgsFeature &feature, const QgsRenderContext &c ) const
+{
+  QSizeF stackedSize( 0, 0 );
+  int enabledDiagramCount = 0;  // We'll add spacing only for enabled subDiagrams
+
+  // Iterate renderers. For each renderer, get the diagram
+  // size for the feature and add it to the total size
+  // accounting for stacked diagram defined spacing
+  for ( const auto &subRenderer : std::as_const( mDiagramRenderers ) )
+  {
+    QSizeF size = subRenderer->sizeMapUnits( feature, c );
+
+    if ( size.isValid() )
+    {
+      enabledDiagramCount++;
+      switch ( mSettings.stackedDiagramMode )
+      {
+        case QgsDiagramSettings::Horizontal:
+          stackedSize.setWidth( stackedSize.width() + size.width() );
+          stackedSize.setHeight( std::max( stackedSize.height(), size.height() ) );
+          break;
+
+        case QgsDiagramSettings::Vertical:
+          stackedSize.setWidth( std::max( stackedSize.width(), size.width() ) );
+          stackedSize.setHeight( stackedSize.height() + size.height() );
+          break;
+      }
+    }
+  }
+
+  if ( stackedSize.isValid() )
+  {
+    const double spacing = c.convertToMapUnits( mSettings.stackedDiagramSpacing(), mSettings.stackedDiagramSpacingUnit(), mSettings.stackedDiagramSpacingMapUnitScale() );
+
+    switch ( mSettings.stackedDiagramMode )
+    {
+      case QgsDiagramSettings::Horizontal:
+        stackedSize.scale( stackedSize.width() + spacing * ( enabledDiagramCount - 1 ), stackedSize.height(), Qt::IgnoreAspectRatio );
+        break;
+
+      case QgsDiagramSettings::Vertical:
+        stackedSize.scale( stackedSize.width(), stackedSize.height() + spacing * ( enabledDiagramCount - 1 ), Qt::IgnoreAspectRatio );
+        break;
+    }
+  }
+  return stackedSize;
+}
+
+void QgsStackedDiagramRenderer::renderDiagram( const QgsFeature &feature, QgsRenderContext &c, QPointF pos, const QgsPropertyCollection &properties ) const
+{
+  if ( !mDiagram )
+  {
+    return;
+  }
+
+  QPointF newPos = pos; // Each subdiagram will have its own newPos
+
+  // Get subrenderers sorted by mode (vertical diagrams are returned backwards)
+  const QList< QgsDiagramRenderer * > stackedRenderers = renderers( true );
+
+  for ( const auto &stackedRenderer : stackedRenderers )
+  {
+    if ( stackedRenderer->rendererName() == QStringLiteral( "Stacked" ) )
```

Comment:
stacked diagrams in stacked diagrams
![inception-kukkendare](https://github.com/user-attachments/assets/426a4eb6-2030-4b0a-a1b1-809822a9f9cb)

Reply:
When I first thought about it, and gave it a try, that specific `if` clause was the only missing piece.

Note this opens the opportunity to have grids of diagrams instead of _only_ vertical or horizontal stacks.

As mentioned in the PR description, this functionality is not yet offered via GUI, but only demonstrated via unit test.